js/modules/loki_app_dot_net_api.js

```diff
@@ -324,24 +411,89 @@ class LokiAppDotNetServerAPI {
         fetchOptions.body = rawBody;
       }
       fetchOptions.headers = new Headers(headers);
-      result = await nodeFetch(url, fetchOptions || undefined);
+
+      if (
+        window.lokiFeatureFlags.useSnodeProxy &&
+        (this.baseServerUrl === 'https://file-dev.lokinet.org' ||
+          this.baseServerUrl === 'https://file.lokinet.org')
+      ) {
+        // make temporary key for this request/response
+        ephemeralKey = libsignal.Curve.generateKeyPair();
+        // mix server pub key with our priv key
+        symKey = libsignal.Curve.calculateAgreement(
+          this.pubKey, // server's pubkey
+          ephemeralKey.privKey // our privkey
+        );
+        // some randomness
+        iv = libsignal.crypto.getRandomBytes(IV_LENGTH);
+
+        log.info('Sending a proxy request to', this.baseServerUrl);
+        result = await this.constructor._sendToProxy(
+          { ...fetchOptions, headers },
+          endpoint,
+          method,
+          { ephemeralKey, symKey, iv }
+        );
+        // log.info('sent to proxy')
+      } else {
+        result = await nodeFetch(url, fetchOptions || undefined);
+      }
     } catch (e) {
-      log.info(`e ${e}`);
+      log.info(
+        'serverRequest nodeFetch/_sendToProxy error:',
+        JSON.stringify(e)
+      );
       return {
         err: e,
       };
     }
     let response = null;
+    let TxtResponse = '';
     try {
-      response = await result.json();
+      TxtResponse = await result.text();
+      response = JSON.parse(TxtResponse);
     } catch (e) {
-      log.warn(`serverRequest json parse ${e}`);
+      log.warn(`serverRequest json parse ${e} ${TxtResponse}`);
       return {
         err: e,
         statusCode: result.status,
       };
     }
 
+    if (
+      window.lokiFeatureFlags.useSnodeProxy &&
+      (this.baseServerUrl === 'https://file-dev.lokinet.org' ||
+        this.baseServerUrl === 'https://file.lokinet.org')
+    ) {
+      // log.info('Got proxy response', response, 'for', method || 'GET', endpoint);
+      if (response.meta && response.meta.code === 200) {
```

Comment:
wouldn't it be better to move the deciphering inside `_sendToProxy` like maxim has node for the snode proxy? Once we don't need it then we can revert back to node fetch without having to clean up the existing code

Reply:
Because of the breaking up the try blocks, it really doesn't work out in a way "we can revert back to node fetch without having to clean up the existing code" See 5ced9a2. Does have some nice advantages; to keep all the crypto inside _sendToProxy and getting rid of that serverRequest/_sendToProxy combine catch failure message that was driving my brain crazy. But not sure if that's really any benefit to make it worth it. IMO It was much easier to remove the feature flag before 5ced9a2.

I can continue 5ced9a2 to move all the decoding inside _sendToProxy. But the whole _sendToProxy probably would make more sense if it lived somewhere else but since nothing else calls it, I wonder if we shouldn't just inline inside of serverRequest.

Reply:
My thoughts for having `_sendToProxy` include both encryption and decryption was so that the logic is grouped together, other functions outside it shouldn't know how a proxy request is structured and thus don't need to do all proxy specific decryption themselves.

inlining the function inside `serverRequest` would just make the function much harder to read so i think having it seperate is good.

@msgmaxim what are your thoughts?

Reply:
I vote for decoding inside `_sendToProxy`. Even if this is the only place we call it, it'd still be better for readability to keep a separate function.

Reply:
ok I'll continue with the refactor